File: Data Structures & Algorithms/longest-turbulent-subarray/submission-5.cpp
class Solution {
public:
    int maxTurbulenceSize(vector<int>& arr) {
        vector<char> rel(arr.size(), ' ');

        for (int i = 1; i < arr.size(); i++) {
            if (arr[i] < arr[i - 1]) {
                rel[i] = 'l';
            } else if (arr[i] > arr[i - 1]) {
                rel[i] = 'g';
            } else {
                rel[i-1] = 'e';
                rel[i] = 'n';
            }
        }

        int curr = 0;
        int max = 0;
        rel[0] = 'n';
        int prev = 'n';


        for (char c : rel) {
            if (c == prev) {
                curr = 0;
            }

            curr++;
            if (curr > max) {
                max = curr;
            }

            if (c == 'e') {
                curr = 0;
            }

            prev = c;
        }

        return max;
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-5.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        # reverse the list from the middle node to the last node
        # now there are two pointers: head and tail
        # to construct new list add the node at head then the node at tail
        # move to the next node for both head and tail
        # repeat until there are no more nodes left to 

        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        curr = slow.next
        slow.next = None

        prev = None
        nextNode = None
        while curr:
            nextNode = curr.next
            curr.next = prev
            prev = curr
            curr = nextNode
        
        tail = prev
        tmpPtr3 = head
        head = head.next

        while tail:
            tmpPtr3.next = tail
            tail = tail.next
            tmpPtr3 = tmpPtr3.next

            if head:
                tmpPtr3.next = head
                head = head.next
                tmpPtr3 = tmpPtr3.next